vars/runTest.groovy

```diff
@@ -80,8 +82,13 @@ Map call(Map config = [:]) {
     }
 
     String cb_result = currentBuild.result
-    int rc = 0
-    rc = sh(script: script, label: flow_name, returnStatus: true)
+    int rc = 2
+    try {
+        sh(script: script, label: flow_name)
+        rc = 0
+    } catch (hudson.AbortException e) {
+        rc = 1
+    }
```

Comment:
What's the benefit of using a `try`/`catch` here vs. how it used to be, given that it means that we lose the return value of test script run in the stash?

Reply:
Here were only checking if the rc value is non-zero.

We get a lot of tickets where the developer filing the ticket only provides the error text from the "error" method called later in the stage.  This makes the step that had the error and potentially useful information highlighted as a place to look, which should make it easier for developers to see where the failure occurred an cut down on the tickets.

Reply:
> Here were only checking if the rc value is non-zero.

But we stash the return code for future stages to analyze.  This change masks the return code with a pre-defined 2, 0 or 1.
 
> We get a lot of tickets where the developer filing the ticket only provides the error text from the "error" method called later in the stage.

Which ultimately is wrong.  Those should surface up as JUnit results.  Users should never have to look at steps to figure out what went wrong.  Similar to [SRE-652](https://daosio.atlassian.net/browse/SRE-652).

Reply:
> Surfacing as Junit results is just one part of the solution.

I don't think so.  It's the entire solution, along with education about the workflow improvements.

> The text of the error message that fails the stage likely needs to be modified to tell people to check the JUnit test results.

At the point that we are surfacing all possible errors/failures into JUnit results, users will be educated that for any/every failure of a PR, the cause will be in test results and that they can completely ignore the step output.
 
> The changing to use try/catch blocks is https://daosio.atlassian.net/browse/SRE-1683

I'm not sure I'm understanding how one is any different from the other.  Is there a "visual" difference in the step reporting of using try/catch vs. the older implementation?  Is there something I am missing there?

To be clear, I understand what SRE-1683 is saying.  I just don't understand how this code change effects that.
 
> It looks like the actual result code can be obtained by parsing one of 'e.toString()' or 'e.getMessage()' which returns "script returned exit code N'

So that could be parsed to stash the real exit value and restore the previous behaviour.  That only seems worth it if the change from the simple `sh()` call to `try`/`catch`.

Reply:
Yes, using a try/catch block causes the "sh" step with a non-zero return to be highlighted in red so that it easily found.

Current push retrieves the rc value from the exception message, which we were stashing to allow its use in post steps so it is there if we ever want to look at specific values.  The default of 255 is for an exception that an exit code could not be parsed from.

This is demonstrated in:
https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-11595/23/pipeline/544

The step that actually ran the test is set to RED to indicate that it was not a SUCCESSFUL test run.
The Unit Test with memcheck on EL 8 step is set to RED as before, but now advises to check the Junit results
The Archive JUnit step is ORANGE for me to indicate that it set the stage to unstable.
The grep for errors in the valgrind memcheck.xml files are RED because they found an issue.
The marking of the stage as unstable because of the valgrind failures is ORANGE for me.

Reply:
> Yes, using a try/catch block causes the "sh" step with a non-zero return to be highlighted in red so that it easily found.

I see.  That's interesting and useful.  But I think we need to be careful not to overdo it, per the comments below.

> This is demonstrated in: https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-11595/23/pipeline/544

I see.  I also see that it didn't stop on the  `UNSTABLE` result from *Unit Tests / Unit Test with memcheck on EL 8*.  Is that `UNSTABLE` stage not setting `currentBuild.currentResult` to `UNSTABLE` and thus triggering https://github.com/daos-stack/pipeline-lib/blob/0afbed05f7b7cd6e283cee05bd890be06e76712c/vars/skipStage.groovy#L187?

> The grep for errors in the valgrind memcheck.xml files are RED because they found an issue.

I don't think this one should be red.  I think there is enough indication in the stage['s steps] that there was a failure without adding more confusing red steps.  Moreover this step is an "internal processing" step that we are adding to more clearly indicate that the over job is unsuccessful, which I think we are doing well enough without making this step red. Users really ought not be interested in the details of this step and making it red I think attracts attention to it where they might even be pasting the output from it into bug reports.

Reply:
Looking at the logs for that run, the Archive JUnit step should have changed the stage status to UNSTABLE from SUCCESSFUL.

There is a check below that which is suppose to report if the stage status was actually changed, and nothing got logged, even though the stage was marked in the UNSTABLE status color.

I need to make some more test diagnostic runs with my test branch to see what is going on.  It is looking like the JUnit report and the "unstable" steps in a post section changed the stage status without changing the currentBuild.result value.

When we first set up the Unit tests, we were doing the JUnit report in the "Steps" section, not the "Post" section, and I wonder if that made a difference in the behavior.

Calling the "error" step instead of the "unstable" step does stop the pipeline though.

Reply:
The reason that the test errors for "Unit Test with memcheck for EL 8" are not stopping the build is that the DAOS jenkinsfile is specifying "ignore_failure" to allow them to continue.

In the current push, the ignore_failure should also control if junit reporting of the detected valgrind issues will cause the stage to be marked as failed or just marked as unstable.

Reply:
So, if the caller requested `ignore_failure: true` and valgrind errors are found we disregard the caller's request and make the stage an `FAILED` but if the caller did not request `ignore_failure: true`, make it `UNSTABLE` and continue on with the job?

That seems a bit unintuitive to me.  Isn't the entire cause of SRE-1729 in fact simply that the [daos Jenkinsfile](https://github.com/daos-stack/daos/blob/master/Jenkinsfile) should be changed to not request [ignore_failure: true](https://github.com/daos-stack/daos/blob/master/Jenkinsfile#L819)?

I think the rest of what's in this PR is great, especially the part about creating a JUnit result when Valgrind failures are found, but I don't think we need to contort the meaning of `ignore_failure: true` uniquely for Valgrind stages when it's more general (and consistent) meaning works for this stage as well as any other.